### PR TITLE
Move collision detection and movements to server side

### PR DESCRIPTION
**Pull request by codymaki**
## Changes

- moved collision detection to server side
- refactored enemy state to be at enemy level to allow for multiple enemies
- updated socket logic to receive asset location data
- removed hitbox display logic from front end, that will need to be refactored to account for new backend code.
- moved animation logic to backend
- refactored player logic and moved physics to backend, renaming from ninja

## References

- https://github.com/codymaki/Bokudos-server/pull/2 - server side code

## Screenshots

**example with two players and a bunch of enemies**
image

### CODE REVIEW

File: src/main/java/com/bokudos/bokudosserver/external/stagebuilder/Tiles.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * This class uses a HashMap, therefore it is not threadsafe.
+ * A different Tiles instance should be used per game even if the tiles are built from the same stage.
+ */
+public class Tiles {
+
+    private Map<Integer, Map<Integer, Boolean>> data;
```
Comment (BadassBison) [inline src/main/java/com/bokudos/bokudosserver/external/stagebuilder/Tiles.java:12]: Nested map o.0

File: src/main/java/com/bokudos/bokudosserver/physics/PhysicsSettings.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    private double msPerTick;
+    private double gravityAcceleration;
+    private double terminalVelocity;
+    private double tickRate;
+    private double jumpSpeed;
+    private double movementSpeed;
+    private double movementAcceleration;
+
+    public PhysicsSettings(double tickRate) {
+        this.tickRate = tickRate;
```
Comment (BadassBison) [inline src/main/java/com/bokudos/bokudosserver/physics/PhysicsSettings.java:21]: I think this is a great idea

File: src/main/java/com/bokudos/bokudosserver/threads/GameThread.java [hunk 3]
```diff
@@ -2,58 +2,95 @@
...
+import com.bokudos.bokudosserver.physics.PlayerPhysics;
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import lombok.extern.slf4j.Slf4j;
 import org.springframework.web.socket.TextMessage;
 import org.springframework.web.socket.WebSocketSession;
 
 import java.io.IOException;
-import java.util.ArrayList;
 import java.util.HashMap;
```
Comment (BadassBison) [inline src/main/java/com/bokudos/bokudosserver/threads/GameThread.java:24]: What's the difference between the hashmap and map?

**Review by BadassBison — APPROVED** [review]
Great work! I'll get into doing some work in this project with you this weekend. I might need you to hold my hand till I get familiar with all the Java stuff I don't fully understand

File: src/main/java/com/bokudos/bokudosserver/threads/GameThread.java (hunk 3 above)
Comment (codymaki) [reply src/main/java/com/bokudos/bokudosserver/threads/GameThread.java:24]: Map is an interface, typically a lot of times in Java people declare maps like this for example:
`Map<String, Object> thisMap = new HashMap<>();`
Where Hashmap is the specific implenation of a Map, but Map is just the interface. The nice thing about declaring Map this way and then referencing them as Map, you can easily change the type of the map only where it is initiallized without changing it elsewhere. The HashMap is propbably the most basic type of Map in Java, there are also Hashtables, ConcurrentHashMaps, SortedMaps, and others that all have their different purposes. Hope that answers your question.

File: src/main/java/com/bokudos/bokudosserver/external/stagebuilder/Tiles.java (hunk 1 above)
Comment (codymaki) [reply src/main/java/com/bokudos/bokudosserver/external/stagebuilder/Tiles.java:12]: haha yeah, we can definitely update this. I have been thinking of better ways to store the tile data.

**Pull request merged**